incremental-font-transfer/src/patch_group.rs

```diff
@@ -142,72 +146,52 @@ impl<'a> PatchGroup<'a> {
-        let mut partial_invalidation_ift: Vec<PartialInvalidationPatch> = vec![];
-        let mut partial_invalidation_iftx: Vec<PartialInvalidationPatch> = vec![];
+        let mut full_invalidation: Vec<CandidatePatch> = vec![];
+        let mut partial_invalidation_ift: Vec<CandidatePatch> = vec![];
+        let mut partial_invalidation_iftx: Vec<CandidatePatch> = vec![];
         // TODO(garretrieger): do we need sorted order, use HashMap instead?
         let mut no_invalidation_ift: BTreeMap<String, NoInvalidationPatch> = Default::default();
         let mut no_invalidation_iftx: BTreeMap<String, NoInvalidationPatch> = Default::default();
 
         // Step 1: sort the candidates into separate lists based on invalidation characteristics.
-        for uri in candidates.into_iter() {
-            // TODO(garretrieger): for efficiency can we delay uri template resolution until we have actually selected patches?
-            // TODO(garretrieger): for btree construction don't recompute the resolved uri, cache inside the patch uri object?
-            match uri.encoding() {
-                PatchEncoding::TableKeyed {
-                    fully_invalidating: true,
-                } => full_invalidation.push(FullInvalidationPatch(uri.into())),
-                PatchEncoding::TableKeyed {
-                    fully_invalidating: false,
-                } => {
-                    if Some(uri.expected_compatibility_id()) == ift_compat_id.as_ref() {
-                        partial_invalidation_ift.push(PartialInvalidationPatch(uri.into()))
-                    } else if Some(uri.expected_compatibility_id()) == iftx_compat_id.as_ref() {
-                        partial_invalidation_iftx.push(PartialInvalidationPatch(uri.into()))
-                    }
-                }
-                PatchEncoding::GlyphKeyed => {
-                    if Some(uri.expected_compatibility_id()) == ift_compat_id.as_ref() {
-                        no_invalidation_ift
-                            .insert(uri.uri_string(), NoInvalidationPatch(uri.into()));
-                    } else if Some(uri.expected_compatibility_id()) == iftx_compat_id.as_ref() {
-                        no_invalidation_iftx
-                            .insert(uri.uri_string(), NoInvalidationPatch(uri.into()));
-                    }
-                }
-            }
-        }
+        Self::group_patches(
+            candidates,
+            ift_compat_id,
+            iftx_compat_id,
+            &mut full_invalidation,
+            &mut partial_invalidation_ift,
+            &mut partial_invalidation_iftx,
+            &mut no_invalidation_ift,
+            &mut no_invalidation_iftx,
+        );
 
         // Step 2 - now make patch selections in priority order: first full invalidation, second partial, lastly none.
-        if let Some(patch) = full_invalidation.into_iter().next() {
+        if let Some(patch) = Self::select_invalidating_candidate(full_invalidation.into_iter()) {
```

Comment:
if the generic param here as `T: IntoIterator` instead of `T: Iterator` you wouldn't need the `into_iter` calls.